feat: Implement a one-dimensional interval class

stephenswat opened the pull request:
This merge request is part one of four in my implementation of an orthogonal range search seed finder. That implementation relies on a _k_-d tree, which relies on _k_-dimensional orthogonal intervals. A _k_-dimensional orthogonal interval is simply the product of _k_ one-dimensional intervals, and as such the one-dimensional interval is the first part in the series.

This new class represents the inclusive range [_min_, _max_], and has several methods for shrinking, expanding, and comparing ranges. I have attempted to make the class quite general, and the scalar type is templated.

I have also implemented tests for this new class, of course!

asalzburger commented on Core/include/Acts/Utilities/Range1D.hpp:
```diff
@@ -0,0 +1,322 @@
+// This file is part of the Acts project.
+//
+// Copyright (C) 2016-2018 CERN for the benefit of the Acts project
```
2021.

Thanks for the renaming!

stephenswat replied:
Ah of course, silly me. Should that be `2016-2021`, `2021-2021`, or `2021`?

asalzburger commented on Core/include/Acts/Utilities/Range1D.hpp:
```diff
@@ -0,0 +1,322 @@
+  ///
+  /// @param min The minimum value in the range (inclusive)
+  /// @param max The maximum value in the range (inclusive)
+  Range1D(Type min, Type max) : m_min(min), m_max(max) {}
```
If you have a chain of Ranges, and both borders are inclusive, isn't that somewhat inconsistent?

Sure, that's just a numerical edge effect, but I was wondering if one could use a [x,y) range definition instead.

stephenswat replied:
Agreed, let's move to a half-open interval definition instead.

asalzburger commented on Core/include/Acts/Utilities/Range1D.hpp:
```diff
@@ -0,0 +1,322 @@
+  /// @return false The ranges are not equal
+  bool operator==(const Range1D<Type>& o) const {
+    return min() == o.min() && max() == o.max();
+  }
```
Had to check this on a piece of paper, approved.

asalzburger reviewed (approved):
Straight forward , there's just the one header to correct and we can put it in.

The inclusion of both boundaries is more of a question how we deal with this in case one wanted to use this Range class for chaining.

stephenswat commented:
In accordance with Andi's review, I've updated the copyright notice, I've made the range semi-open, and I've added some tests to make sure this holds.

asalzburger commented:
Let's wait for the CI but then this is ready to go.

stephenswat commented:
Looks like the CI has passed, it just needs a new approval at this point. :slightly_smiling_face:
